### PR TITLE
Items new

### LINKED ISSUE #6: Interactables
- [X] Create `Interactable` component that fires a `UnityEvent` when in contact with the player character (#15)
- [ ] Create oxygen (skeleton by #15)
- [ ] Create caffeine (skeleton by #15)
- [X] Place items (#19)

Here are the steps to implement the functionalities of Oxygen and Caffeine.

1. In `CombatEntity` component, create public methods to increase current health, attack, and defense.
   - health: `InternalCurrentHealth = currentHealth + <value>;`
   - attack: `InternalCurrentAttack = currentAttack + <value>;`
   - defense: `InternalCurrentDefense = currentDefense + <value>;`
1. In the `OnConsume` methods of `Oxygen` and `Caffeince` monobehaviours, call the appropriate methods created above by first getting the `CombatEntity` component.
   ```csharp
   public void OnConsume(GameObject player)
   {
       CombatEntity combatEntity = player.GetComponent<CombatEntity>();
       // do stuff with combatEntity
   }
   ```

### PR DESCRIPTION
**Pull request by ericsrna**
Implemented Oxygen and Caffeine function, and create New branch of Items because of previous Items Branch has some merging issues.

### CODE REVIEW

File: Assets/Scripts/Combat/CombatEntity.cs [hunk 1]
```diff
@@ -122,6 +122,25 @@ public void GetAttacked(int rawDamage)
             InternalCurrentHealth = currentHealth - (rawDamage - currentDefense);
         }
 
+        public void GetDecreased(int rawDamage)
+        {
+            InternalCurrentHealth = currentHealth - rawDamage;
+        }
+
+        public void GetPowerUp(string type, int value)
```
Comment (JameeKim) [inline Assets/Scripts/Combat/CombatEntity.cs:130]: It seems like you wanted a versatile method to handle all the stats. While this gives us flexibility, it is also an error-prone approach. And this is where I'm kind of paranoid, and I always try to prevent possible human errors whenever possible.

The two choices I can currently think of are creating separate methods for separate things--so making them do one thing at a time--and limit what can be put into the input argument. To elaborate on the second option, there is a nice feature you can use called `enum`. I believe you've already heard of it, and might have actually used it yourself. (Probably not in C#, but a lot of programming languages have enums, so yeah.)

When you declare an enum and use it as a public field of a component, Unity actually recognizes it and in the inspector, lets you choose the value from a dropdown of all declared values of that enum for the field. Also if you use enums, you can get a compile error when you do not put in the right type of argument when calling the method in the code. I think this is better than using just strings in the aspect of preventing human errors like typos.

That being said, I personally feel like just having separate methods could be better here because we're not creating a huge game and there is very little possibility that more types of stats will be added. But I suppose this is more of a choice of style preferences, so the final decision is yours to make. This was just my thoughts that you can consider, but I'm not enforcing you to follow the things I said above.

Edit: typos

File: Assets/Scripts/Items/Oxygen.cs [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
+        public int powerUpAmount = 10;
+        string typeOfItem = "Oxygen";
+
+        private CombatEntity combatEntity;
+
         public void OnConsume(GameObject player)
         {
-            Debug.Log("Player consumed an oxygen crystal");
-            // TODO
+            combatEntity = player.GetComponent<CombatEntity>();
```
Comment (JameeKim) [inline Assets/Scripts/Items/Oxygen.cs:17]: ~Same as the `Caffeine` component, it would be better if you call `GetComponent` only once. It is trivial in our case, but just common practice.~

Nevermind, it is the component of the player. My bad. Nothing wrong here.

File: Assets/Scripts/Items/Caffeine.cs [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+        public float powerUpDuration = 5.0f;
+        string typeOfItem = "Caffeine";
+
+        private CombatEntity combatEntity;
+
         public void OnConsume(GameObject player)
         {
-            Debug.Log("Player consumed a caffeine");
-            // TODO
+            combatEntity = player.GetComponent<CombatEntity>();
```
Comment (JameeKim) [inline Assets/Scripts/Items/Caffeine.cs:19]: ~It is probably better to cache the reference to the `CombatEntity` component only once (in `Start` event function).~

Nevermind, I was thinking of something else. My bad.

File: Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs [hunk 4]
```diff
@@ -27,8 +42,33 @@ private void FixedUpdate()
             float horiz = Input.GetAxisRaw("Horizontal");
             float vert = Input.GetAxisRaw("Vertical");
 
-            Vector2 direction= new Vector2(horiz, vert).normalized;
+            Vector2 direction = new Vector2(horiz, vert).normalized;
             rigidBody.AddForce(direction * acceleration);
+
+            //controller.DisableInputForSeconds(paralyzeDuration); // disable input for stalling effect
+
+            PowerUpTimer(); // Consuming Caffeine adds time duration and runs its timer
```
Comment (JameeKim) [inline Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs:50]: This has nothing to do with the physics simulation, so generally you would call this in `Update` event.

File: Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs [hunk 5]
```diff
@@ -27,8 +42,33 @@ private void FixedUpdate()
...
+        }
+                
+        public void PowerUpDuration(int amount, float duration)
+        {
+            this.powerUpValue += amount;
+            this.powerUpTimer += duration;
+        }
+
+        private void PowerUpTimer()
```
Comment (JameeKim) [inline Assets/Scripts/PlayerCharacter/PlayerCharacterController.cs:59]: This is a separate thing to controlling the movement of player character, which is what `PlayerCharacterController` is trying to do. So it would be better if this gets its own component, something like `CaffeineEffectTracker` or so.

**Review by JameeKim — CHANGES_REQUESTED** [review]
Thank you! Here are some points I want to make.

1. I think you forgot to switch to the master branch before creating a new branch to implement the functions of the items. As a result, this pull request also has the changelist from the previous request you made, with decreasing the player health. While it is possible to fix this in this pull request, I think it could be easier if you make a fresh pull request from a new branch created from the right base branch (the master branch).
1. Ok, so it seems like there was a miscommunication here. I thought the caffeine item would increase the stats permanantly, but you thought the effects would be temporary, I think? It doesn't really matter, but maybe making the stat increase permanant would be easier to implement. For example, how are we going to indicate how many caffeine effects are currently in effect at the time, and how about the time left? I'll leave the final decision to you, and definitely work together on things that you want to work together.
1. As mentioned in the specific part of the code below, it is your choice of preference whether to make a single method to handle stat changes or to create separate ones for each. But I would highly recommend making an `enum` if you want to have a single method.

To summarize,

1. I suggest creating a new pull request with a new branch from the master branch in order to get rid of the part of the health decreasing system from this change.
1. Make the caffeine do its effect temporarily or permanantly, and feel free to reach for help if you need any.
1. Make separate methods for the stat changes or use an `enum` if you want to use a single method.

Again, thank you for your work! I hope this feedback isn't too overwhelming.

**Comment by JameeKim** [comment]
A new PR #46 is made in place of this PR. Closing.
